git/gogit/client.go

```diff
@@ -254,9 +254,12 @@ func (g *Client) validateUrl(u string) error {
 		return errors.New("URL cannot contain credentials when using HTTP")
 	}
 
-	if httpOrEmpty && g.authOpts != nil &&
-		(g.authOpts.Username != "" || g.authOpts.Password != "") {
-		return errors.New("basic auth cannot be sent over HTTP")
+	if httpOrEmpty && g.authOpts != nil {
+		if g.authOpts.Username != "" || g.authOpts.Password != "" {
+			return errors.New("basic auth cannot be sent over HTTP")
+		} else if g.authOpts.BearerToken != "" {
+			return errors.New("bearer token cannot be sent over HTTP")
+		}
```

Comment:
I think we should have the bearer token take precedence and have this behavior properly documented.

Reply:
I have a quick question from looking at the code in `NewAuthOptions()`. It sets username and password from url as well. Should that part be left alone so it only affects the preference of bearer token from a Secret?

Reply:
I guess the question is similar with `Client.validateUrl()`.
Should it fail if both `authOpts.BearerToken` and `ru.User` is set? 
Or only if `authOpts.BearerToken` and `authOpts.Username`(and/or password) is set?

Reply:
Regarding`NewAuthOptions()`, I think this is the most readable way to represent the requirement, at the end, but before `opts.Validate()`:

```
	// Prefer bearer token over basic auth if both are provided
	if opts.BearerToken != "" && (opts.Transport == HTTP || opts.Transport == HTTPS) {
		opts.Username = ""
		opts.Password = ""
	}
```

What do you think? This will reset username and password whether it's from a Secret or URL.

Reply:
> > Or only if authOpts.BearerToken and authOpts.Username(and/or password) is set?
> 
> Yes, I believe we should fail for this case only.
> 
> Please correct me if it doesn't sound right. I may be overlooking something.

I guess that makes the most sense, as the user from the URL is no longer used for anything at this point, so it's only the chosen auth options that we want to validate. In the normal case it won't ever trigger if the code I suggested above is added, but as you mention with someone skipping `NewAuthOptions()` it could trigger.

Reply:
Am I understanding the requirement for `validateUrl()` to be like this? Added above the check for `credentialsOverHTTP` since this is relevant in either case.

```go
if g.authOpts != nil {
	httpOrHttps := g.authOpts.Transport == git.HTTP || g.authOpts.Transport == git.HTTPS
	hasUsernameOrPassword := g.authOpts.Username != "" || g.authOpts.Password != ""
	hasBearerToken := g.authOpts.BearerToken != ""

	if httpOrHttps && hasBearerToken && hasUsernameOrPassword {
		return errors.New("basic auth and bearer token cannot be set at the same time")
	}
}
```

Reply:
Thank you! The code looks much clearer now. I was worried about changing too much and breaking something.

But I think there is a new bug here. We use `caFile` when we connect to a GitLab instance with self signed certificates over https, otherwise Flux will complain about invalid certificate.

From the documentation it looks like that field is meant for https and not ssh:

> HTTPS Certificate Authority
> To provide a Certificate Authority to trust while connecting with a Git repository over HTTPS, the referenced Secret can contain a .data.caFile value.

https://fluxcd.io/flux/components/source/gitrepositories/#https-certificate-authority

Reply:
Looks like moving the basic auth credentials from the URL behind `if len(data) > 0` makes these scenarios fail:

```go
{
	name: "Sets username and password from URL",
	URL:  "http://someuser:somepass@example.com",
	data: nil,
	wantFunc: func(g *WithT, opts *AuthOptions) {
		g.Expect(opts.Username).To(Equal("someuser"))
		g.Expect(opts.Password).To(Equal("somepass"))
	},
},
{
	name: "Sets username and password from URL with empty map",
	URL:  "http://someuser:somepass@example.com",
	data: map[string][]byte{},
	wantFunc: func(g *WithT, opts *AuthOptions) {
		g.Expect(opts.Username).To(Equal("someuser"))
		g.Expect(opts.Password).To(Equal("somepass"))
	},
},
```

This works though, with just having a secret with something in it:

```go
{
	name: "Sets username and password from URL with map with content",
	URL:  "http://someuser:somepass@example.com",
	data: map[string][]byte{
		"random": []byte("value"),
	},
	wantFunc: func(g *WithT, opts *AuthOptions) {
		g.Expect(opts.Username).To(Equal("someuser"))
		g.Expect(opts.Password).To(Equal("somepass"))
	},
},
```

All these would work previously, so I assume it should continue to work?

Reply:
I pushed an update to `NewAuthOptions()` now, with changes to handle these issues. Let me know if it looks correct.

Reply:
Great! Note that I haven't pushed the discussed changes to `validateUrl()` yet. I'm going to write the required tests shortly.

Reply:
The last changes have been pushed now.

Reply:
Good thing there was a test for that! Pushed an update now.

Reply:
I've run all the tests here and in `source-controller` now and it looks good.

Also tested `source-controller` in my cluster with these gitrepositories:

* GitLab with https (self signed) and basic auth with caFile
* Bitbucket  with https and bearer token
* GitHub with ssh

Reply:
Thank you, it's been great working with you too 🙂